### PR TITLE
managers for services

### PR DESCRIPTION
**Pull request by altendky**
### Purpose:

Trying to move towards context managers so, well, we can use more context managers.

### Draft For:
- [x] https://github.com/Chia-Network/chia-blockchain/pull/16710

### CODE REVIEW

**Comment by emlowe** [comment]
I like this and it seems to simplify some code -

File: chia/util/misc.py [hunk 1]
```diff
@@ -280,3 +282,95 @@ def setup_async_signal_handler(self, handler: AsyncHandler) -> None:
         self.setup_sync_signal_handler(
             handler=functools.partial(self.threadsafe_sync_signal_handler_for_async, handler=handler)
         )
+
+
+@dataclass
+class SplitManager(Generic[T]):
```
Comment (arvidn) [inline chia/util/misc.py:288]: what is "SplitManager" used for? the name doesn't make that immediately obvious to me

**Comment by arvidn** [comment]
looks reasonable

File: chia/util/misc.py (hunk 1 above)
Comment (altendky) [reply chia/util/misc.py:288]: It is used in some tests.  I think they should probably be refactored so we can delete this.  Maybe I should have done that instead of this.  I dunno.

File: chia/server/start_service.py [hunk 2]
```diff
@@ -175,68 +189,97 @@ async def _connect_peers_task_handler(self) -> None:
...
+                self._log.info("Waiting for ChiaServer to be closed")
+                await self._server.await_closed()
+
+                if self.rpc_server:
+                    self._log.info("Waiting for RPC server")
+                    await self.rpc_server.await_closed()
+                    self._log.info("Closed RPC server")
+
+        # this is a blocking call, waiting for the UPnP thread to exit
+        self.upnp.shutdown()
```
[the remark this answers is not in the document]
Comment (altendky) [reply chia/server/start_service.py:280]:
```
# NOTE: avoid start=False, this is presently used for corner case setup type tests
```

`start` being `False` is in support of present test patterns and the behavior should go away and not be leveraged.  It allows the service to actually be started later but still have the context manager make sure it gets shut down.

File: chia/server/start_service.py [hunk 3]
```diff
@@ -175,68 +189,97 @@ async def _connect_peers_task_handler(self) -> None:
                         resolved_peers[unresolved] = resolved_new
             await asyncio.sleep(self.reconnect_retry_seconds)
 
-    async def start(self) -> None:
-        # TODO: move those parameters to `__init__`
-        if self._did_start:
-            return None
+    async def run(self) -> None:
+        try:
+            with Lockfile.create(service_launch_lock_path(self.root_path, self._service_name), timeout=1):
```
[the remark this answers is not in the document]
Comment (altendky) [reply chia/server/start_service.py:194]: GitHub is not aligning that bit of the diff great.  If you look at the existing `.run()` method, the lock file is already in use there.

File: chia/introducer/introducer.py [hunk 4]
```diff
@@ -35,6 +36,15 @@ def __init__(self, max_peers_to_send: int, recent_peer_threshold: int):
         self._server: Optional[ChiaServer] = None
         self.log = logging.getLogger(__name__)
 
+    @contextlib.asynccontextmanager
+    async def manage(self) -> AsyncIterator[None]:
+        await self._start()
+        try:
+            yield
+        finally:
+            self._close()
```
[the remark this answers is not in the document]
Comment (altendky) [reply chia/introducer/introducer.py:45]: That is an interesting point.  This idea could relate to slower test shutdown.  I had not thought about this in detail.  How concerned are you?  I'm primarily concerned about the getting stuff shut down correctly to avoid all the issues with that.  But maybe we could find a good solution that retains the safety but allows premature processing for efficiency in needed cases.

File: chia/server/start_service.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/server/start_service.py:280]: https://github.com/chia-network/chia-blockchain/commit/6575d14c32a97488d7aeebd202132bcd2f302481

[the remark this answers is not in the document]
Comment (altendky) [reply chia/server/start_service.py:280]: I think we could easily reduce the indentation as you say, but I don't think it would make the code actually more clear.  Hopefully we can just get around to getting rid of the `start` control entirely.

File: chia/util/misc.py (hunk 1 above)
Comment (altendky) [reply chia/util/misc.py:288]: ...  it is used to allow separation of the entry and exit of a context manager in ways that they aren't meant to be used.  This is related to what we had before with separate start and stop methods.  Note the warning `# NOTE: only for transitional testing use, please avoid usage`.  Basically, I'm not fixing all the testing patterns that depended on the 'unsafe' separation of start and stop.  They can be addressed in another PR.

File: chia/introducer/introducer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arvidn) [reply chia/introducer/introducer.py:45]: > How concerned are you?
not very. We're struggling with correctness primarily.

**Review by arvidn — DISMISSED** [review]

**Review by arvidn — APPROVED** [review]
